tests/test_username_api.py

```diff
@@ -0,0 +1,249 @@
+import unittest
+import json
+import yaml
+import string
+import random
+import pytest
+import username_api
+
+data = yaml.load(open('tests/test_data.yml'))
+websites = yaml.load(open('websites.yml'))
+
+invalid_username = '$very%long{invalid}user(name)'
+
+def generate_random_username(length):
+  return ''.join(random.choice(string.ascii_uppercase + string.ascii_lowercase \
+    + string.digits) for _ in range(length))
+
+def assert_response(app, website, user, status):
+  expected = get_expected_response(website, user, status)
+  actual = get_response(app, website, user)
+  assert expected == actual
+
+def get_response(app, website, user):
+  resp = app.get('/check/{}/{}'.format(website, user))
+  return json.loads(resp.get_data())
+
+def get_expected_response(website, user, status):
+  return {
+    'possible': True,
+    'status': status,
+    'url': websites['urls'].get(website, 'https://{w}.com/{u}').format(
+      w=website,
+      u=user
+    )
+  }
+
+class TestUsernameApi(object):
+  @classmethod
+  def setup_class(self):
+    self.app = username_api.app.test_client()
+    self.data = data
+
+  @pytest.mark.parametrize('user', data['github']['taken_usernames'])
+  def test_taken_github_username(self, user):
+    assert_response(self.app, 'github', user, 200)
+
+  @pytest.mark.parametrize('user', data['soundcloud']['taken_usernames'])
+  def test_taken_soundcloud_username(self, user):
+    assert_response(self.app, 'soundcloud', user, 200)
+
+  @pytest.mark.parametrize('user', data['gitlab']['taken_usernames'])
+  def test_taken_gitlab_username(self, user):
+    assert_response(self.app, 'gitlab', user, 200)
+
+  @pytest.mark.parametrize('user', data['tumblr']['taken_usernames'])
+  def test_taken_tumblr_username(self, user):
+    assert_response(self.app, 'tumblr', user, 200)
+
+  @pytest.mark.parametrize('user', data['behance']['taken_usernames'])
+  def test_taken_behance_username(self, user):
+    assert_response(self.app, 'behance', user, 200)
+
+  @pytest.mark.parametrize('user', data['pinterest']['taken_usernames'])
+  def test_taken_pinterest_username(self, user):
+    assert_response(self.app, 'pinterest', user, 200)
+
+  @pytest.mark.parametrize('user', data['instagram']['taken_usernames'])
+  def test_taken_instagram_username(self, user):
+    assert_response(self.app, 'instagram', user, 200)
+
+  @pytest.mark.parametrize('user', data['twitter']['taken_usernames'])
+  def test_taken_twitter_username(self, user):
+    assert_response(self.app, 'twitter', user, 200)
+
+  @pytest.mark.parametrize('user', data['twitter']['available_usernames'])
+  def test_available_github_username(self, user):
+    expected = get_expected_response('github', user, 404)
+    actual = get_response(self.app, 'github', user)
+
+    if expected == actual:
+      assert expected == actual
```

Comment:
this should be always run at the bottom of the method, as a human needs to update the `'available_usernames'` data if it fails

add custom fail messages, letting the reader know

1. if a random username also failed to be detected as available.
2. what random username they can use instead of the taken username.